bind_uploader.go

```diff
@@ -3,10 +3,256 @@
+				numBytes += len(bindingRows[rowNum])
+				rowNum++
+			}
+			// concatenate all byte arrays into 1 and put into input stream
+			var b bytes.Buffer
+			b.Grow(numBytes)
+			for i := startIdx; i < rowNum; i++ {
+				b.Write(bindingRows[i])
+			}
+
+			bu.fileCount++
+			filename := strconv.Itoa(bu.fileCount)
+			data, err = bu.uploadStreamInternal(&b, filename, true)
+			if err != nil {
+				return nil, err
+			}
+			startIdx = rowNum
+			numBytes = 0
+		}
+		return data, nil
+	}
+	return nil, nil
+}
+
+func (bu *bindUploader) uploadStreamInternal(inputStream *bytes.Buffer, dstFilename string, compressData bool) (*execResponseData, error) {
+	err := bu.createStageIfNeeded()
+	if err != nil {
+		return nil, err
+	}
+	stageName := bu.stagePath
+	if stageName == "" {
+		return nil, &SnowflakeError{
+			Number:  ErrBindUpload,
+			Message: "stage name is null",
+		}
+	}
+	if dstFilename == "" {
+		return nil, &SnowflakeError{
+			Number:  ErrBindUpload,
+			Message: "destination file is null",
+		}
+	}
+
+	var putCommand strings.Builder
+	// use a placeholder for source file
+	putCommand.WriteString("put file:///tmp/placeholder ")
+	// add stage name surrounded by quotations in case special chars are used in directory name
+	putCommand.WriteString("'")
+	putCommand.WriteString(stageName)
+	putCommand.WriteString("'")
+	putCommand.WriteString(" overwrite=true")
+	data, err := bu.sc.exec(bu.ctx, putCommand.String(), false, false, false, []driver.NamedValue{})
+	if err != nil {
+		return nil, err
+	}
+
+	sfa := &snowflakeFileTransferAgent{
+		data:                       data.Data,
+		command:                    putCommand.String(),
+		sourceStream:               inputStream,
+		dstFileNameForStreamSource: dstFilename,
+		compressSourceFromStream:   compressData,
+	}
+	sfa.execute()
+	return sfa.result()
+}
+
+func (bu *bindUploader) createStageIfNeeded() error {
+	if bu.arrayBindStage != "" {
+		return nil
+	}
+	data, err := bu.sc.exec(bu.ctx, createStageStmt, false, false, false, []driver.NamedValue{})
+	if !data.Success {
+		code, err := strconv.Atoi(data.Code)
+		if err != nil {
+			return err
+		}
+		return &SnowflakeError{
+			Number:   code,
+			SQLState: data.Data.SQLState,
+			Message:  err.Error(),
+			QueryID:  data.Data.QueryID}
+	}
+	if err != nil {
+		return err
+	}
+	bu.arrayBindStage = stageName
+	if err != nil {
+		newThreshold := "0"
+		bu.sc.cfg.Params[sessionArrayBindStageThreshold] = &newThreshold
+	}
+	return nil
+}
+
+// transpose the columns to rows and write them to a list of bytes
+func (bu *bindUploader) buildRowsAsBytes(columns []driver.NamedValue) ([][]byte, error) {
+	numColumns := len(columns)
+	if columns[0].Value == nil {
+		return nil, &SnowflakeError{
+			Number:  ErrBindSerialization,
+			Message: "no binds found in the first column",
+		}
+	}
+
+	_, arr := snowflakeArrayToString(&columns[0])
+	numRows := len(arr)
+	for i := 0; i < numColumns; i++ {
+		_, arr = snowflakeArrayToString(&columns[i])
```

Comment:
should  we save 'arr' returned snowflakeArrayToString to avoid doing it again in loop at line 169? Also if we have done it for index 0, this loop on line 152 should start from index 1.

Reply:
Fixed.

Reply:
saving arr to avoid call snowflakeArrayToString again in line 170?